File: backend/app/schemas/lab.py
"""
Lab schemas — request/response models for lab API.
"""
from datetime import datetime
from typing import Optional
from pydantic import BaseModel, ConfigDict


# --- Response schemas ---

class LabResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    subject_id: int
    textbook_id: Optional[int] = None
    title: str
    description: Optional[str] = None
    lab_type: str
    config: dict = {}
    is_active: bool = True
    thumbnail_url: Optional[str] = None


class LabProgressResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    lab_id: int
    progress_data: dict = {}
    xp_earned: int = 0
    completed_at: Optional[datetime] = None


class LabTaskResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    lab_id: int
    title: str
    task_type: str
    task_data: dict = {}
    xp_reward: int = 10
    order_index: int = 0


class LabTaskAnswerResult(BaseModel):
    is_correct: bool
    explanation: Optional[str] = None
    xp_earned: int = 0


# --- Request schemas ---

class ProgressUpdateRequest(BaseModel):
    progress_data: dict = {}


class TaskAnswerRequest(BaseModel):
    answer_data: dict = {}

File: lab-app/src/types/lab.ts
export interface Lab {
  id: number;
  subject_id: number;
  textbook_id: number | null;
  title: string;
  description: string | null;
  lab_type: 'map' | 'molecule_3d' | 'simulation' | 'anatomy';
  config: Record<string, unknown>;
  is_active: boolean;
  thumbnail_url: string | null;
}

export interface LabProgress {
  id: number;
  lab_id: number;
  progress_data: Record<string, unknown>;
  xp_earned: number;
  completed_at: string | null;
}

export interface LabTask {
  id: number;
  lab_id: number;
  title: string;
  task_type: 'find_on_map' | 'order_events' | 'choose_epoch' | 'quiz';
  task_data: Record<string, unknown>;
  xp_reward: number;
  order_index: number;
}

export interface LabTaskAnswerResponse {
  is_correct: boolean;
  explanation: string | null;
  xp_earned: number;
}

export interface EpochData {
  id: string;
  name: string;
  period: string;
  start_year: number;
  end_year: number;
  description: string;
  paragraph_id: number | null;
  color: string;
}
